src/engine/Service.cpp

```diff
@@ -237,48 +275,33 @@ std::optional<std::string> Service::getSiblingValuesClause() const {
 }
 
 // ____________________________________________________________________________
-template <size_t I>
-void Service::writeTsvResult(cppcoro::generator<std::string_view> tsvResult,
-                             IdTable* idTablePtr, LocalVocab* localVocab) {
-  IdTableStatic<I> idTable = std::move(*idTablePtr).toStatic<I>();
-  checkCancellation();
-  size_t rowIdx = 0;
-  std::vector<size_t> numLocalVocabPerColumn(idTable.numColumns());
-  std::string lastLine;
-  const size_t numVariables = parsedServiceClause_.visibleVariables_.size();
-  for (std::string_view line : tsvResult) {
-    // Print first line.
-    if (rowIdx == 0) {
-      LOG(INFO) << "First non-header row of TSV result: " << line << std::endl;
-    }
-    std::vector<std::string_view> valueStrings = absl::StrSplit(line, "\t");
-    if (valueStrings.size() != numVariables) {
-      throw std::runtime_error(absl::StrCat(
-          "Number of columns in row ", rowIdx + 1, " of TSV result is ",
-          valueStrings.size(), " but number of variables in header row is ",
-          numVariables, ". Line: ", line));
-    }
-    idTable.emplace_back();
-    for (size_t colIdx = 0; colIdx < valueStrings.size(); colIdx++) {
-      TripleComponent tc = TurtleStringParser<TokenizerCtre>::parseTripleObject(
-          valueStrings[colIdx]);
-      Id id = std::move(tc).toValueId(getIndex().getVocab(), *localVocab);
-      idTable(rowIdx, colIdx) = id;
-      if (id.getDatatype() == Datatype::LocalVocabIndex) {
-        ++numLocalVocabPerColumn[colIdx];
-      }
-    }
-    rowIdx++;
-    checkCancellation();
-    lastLine = line;
+TripleComponent Service::bindingToTripleComponent(const nlohmann::json& cell) {
+  if (!cell.contains("type") || !cell.contains("value")) {
+    throw std::runtime_error("Missing type or value field in binding.");
   }
-  if (idTable.size() > 1) {
-    LOG(INFO) << "Last non-header row of TSV result: " << lastLine << std::endl;
+
+  const auto type = cell["type"].get<std::string_view>();
+  const auto value = cell["value"].get<std::string_view>();
+
+  TripleComponent tc;
+  if (type == "literal") {
+    if (cell.contains("datatype")) {
+      tc = TurtleParser<TokenizerCtre>::literalAndDatatypeToTripleComponent(
+          value, TripleComponent::Iri::fromIrirefWithoutBrackets(
+                     cell["datatype"].get<std::string_view>()));
+    } else if (cell.contains("xml:lang")) {
+      tc = TripleComponent::Literal::literalWithoutQuotes(
+          value, cell["xml:lang"].get<std::string>());
+    } else {
+      tc = TripleComponent::Literal::literalWithoutQuotes(value);
+    }
+  } else if (type == "uri") {
+    tc = TripleComponent::Iri::fromIrirefWithoutBrackets(value);
+  } else if (type == "blank") {
+    throw std::runtime_error(
+        "Blank nodes in the result of a SERVICE are currently not supported.");
```

Comment:
```suggestion
        "Blank nodes in the result of a SERVICE are currently not supported. For now, consider filtering them out using the ISBLANK function or converting them via the STR function");
```

Maybe even better: Throw a dedicated custom exception here,
catch it in the outer code, and add the information, which variable contained the blank node.